debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/converters/sink/SinkConverter.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.mongodb.converters.sink;
+
+import java.util.Map;
+
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.errors.DataException;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.bson.BsonDocument;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.debezium.connector.mongodb.converters.sink.LazyBsonDocument.Type;
+import io.debezium.table.ColumnNamingStrategy;
+
+public class SinkConverter {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SinkConverter.class);
+
+    private static final SinkRecordConverter SCHEMA_RECORD_CONVERTER = new SchemaRecordConverter();
+    private static final SinkRecordConverter MAP_RECORD_CONVERTER = new MapRecordConverter();
+    private static final SinkRecordConverter JSON_STRING_RECORD_CONVERTER = new JsonStringRecordConverter();
+    private static final SinkRecordConverter BYTE_ARRAY_RECORD_CONVERTER = new ByteArrayRecordConverter();
+
+    public SinkDocument convert(final SinkRecord record, ColumnNamingStrategy columnNamingStrategy) {
+        LOGGER.debug("record: {}", record);
+
+        BsonDocument keyDoc = null;
+        if (record.key() != null) {
+            keyDoc = new LazyBsonDocument(
+                    record,
+                    Type.KEY,
+                    (schema, data) -> getConverter(schema, data).convert(schema, data));
+        }
+
+        BsonDocument valueDoc = null;
+        if (record.value() != null) {
+            valueDoc = new LazyBsonDocument(
+                    record,
+                    Type.VALUE,
+                    (Schema schema, Object data) -> getConverter(schema, data).convert(schema, data));
+        }
+
+        return new SinkDocument(keyDoc, valueDoc);
+    }
+
+    private SinkRecordConverter getConverter(final Schema schema, final Object data) {
+        if (schema != null && data instanceof Struct) {
+            LOGGER.debug("Using schema-ed converter");
+            return SCHEMA_RECORD_CONVERTER;
+        }
+
+        // structured JSON without schema
+        if (data instanceof Map) {
+            LOGGER.debug("Using schemaless / map converter");
+            return MAP_RECORD_CONVERTER;
+        }
+
+        // JSON string
+        if (data instanceof String) {
+            LOGGER.debug("Using JSON string converter");
+            return JSON_STRING_RECORD_CONVERTER;
+        }
+
+        // BSON bytes
+        if (data instanceof byte[]) {
+            LOGGER.debug("Using BSON converter");
+            return BYTE_ARRAY_RECORD_CONVERTER;
+        }
```

Comment:
Why not make use of `if/elseif/else` semantics here?

Reply:
Here I will object... Either use switch or leave it like this. I't definitively more readable IMHO. Once we switch to java 17 switch is likely to be better anyway.

Reply:
As there is immediate return  and the compile should handle it its own way anyway, I think I'll keep it like that until there's another call from @jpechane .

Reply:
@rk3rn3r Keep it as is. We are on Java 17 so might bet better to handle it with pattern matching as thi final solution.
Could you please create a Jira where you record all instances in the connector that would benefit from Java 17 code refactoring?

Reply:
done. filed https://issues.redhat.com/browse/DBZ-8006 for it